### ISSUE TITLE
It doesn't seem helpful for REST error message to include our code

### ISSUE THREAD

**Issue by georgeajit** [marklogic/java-client-api#334]
```
SPARQLQueryManager sparqlQmgr = client.newSPARQLQueryManager();
StringBuffer sparqlQuery = new StringBuffer().append("CREATE GRAPH <DOMICLE>;");
sparqlQuery.append(newline);        

SPARQLQueryDefinition qdef = sparqlQmgr.newQueryDefinition(sparqlQuery.toString());
// Create an empty graph.
sparqlQmgr.executeUpdate(qdef);
```

Running the above snippet second time produces the following stack trace. The JAVA API layer can produce a more meaningful message, rather than XQuery code.

```
com.marklogic.client.FailedRequestException: Local message: failed to apply resource at /graphs/sparql: Internal Server Error. Server Message: XDMP-SPQLGRAPHEXIST: xdmp:security-assert("http://marklogic.com/xdmp/privileges/rest-reader", "execute"), let $rule := conf:get-sparql-protocol-rule() let $params := rest:process-request($rule) let $headers := eput:get-request-headers() let $method := eput:get-request-method($headers) let $env := map:map() let $params := local:validate-params($rule, $env, $params) let $body := switch ($method) case "GET" return text { fn:head((map:get($params, "query"), map:get($params, "update"))) } case "POST" return xdmp:get-request-body(eput:get-content-format($headers, $params))/node() default return fn:error((), "REST-UNSUPPORTEDMETHOD", $method) let $result := semmod:sparql-query($headers, $params, $body) let $response := semmod:results-payload($headers, $params, $result, local:sparql-callback#2) return if ($response instance of node() and $response/self::semmod:malformed-query) then (xdmp:set-response-code(400, "Malformed Query"), $response) else $response -- Graph already exists: DOMICLE . See the MarkLogic server error log for further detail.
    at com.marklogic.client.impl.JerseyServices.checkStatus(JerseyServices.java:4596)
    at com.marklogic.client.impl.JerseyServices.postResource(JerseyServices.java:3492)
    at com.marklogic.client.impl.JerseyServices.executeSparql(JerseyServices.java:5553)
    at com.marklogic.client.impl.SPARQLQueryManagerImpl.executeQueryImpl(SPARQLQueryManagerImpl.java:88)
    at com.marklogic.client.impl.SPARQLQueryManagerImpl.executeQueryImpl(SPARQLQueryManagerImpl.java:81)
    at com.marklogic.client.impl.SPARQLQueryManagerImpl.executeUpdate(SPARQLQueryManagerImpl.java:134)
    at com.marklogic.client.functionaltest.TestBasicSemanticsSparql.testexecuteUpdateCreateSilent(TestBasicSemanticsSparql.java:797)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
    at java.lang.reflect.Method.invoke(Unknown Source)
    at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
    at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
    at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
    at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
    at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
    at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
    at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
    at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
    at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
    at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
    at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
    at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
    at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
    at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
    at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
    at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:50)
    at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
    at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:459)
    at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:675)
    at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:382)
    at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:192)
```

**Comment by grechaw** [comment]
Hi @georgeajit , I fixed this in the REST layer just now.  I'm not sure whether more needs to happen in java code.  Bugtrack did odd things when I cloned the issue there, we'll see how this plays out.

**Comment by georgeajit** [comment]
Charles, I am assigning this back to you since the behaviour has not changed. Not sure if Java Client layer needs to do something additional.

Cloned Java Client API and built the latest jar (08/21/2015) and run the tests again on MarkLogic Server Build 8.0-20150821. The message returned remains the same. Looks like we need to have some filtering to be done in different API layers.

Here is the exception message output:
com.marklogic.client.FailedRequestException: Local message: failed to apply resource at /graphs/sparql: Bad Request. Server Message: XDMP-SPQLGRAPHEXIST: xdmp:security-assert("http://marklogic.com/xdmp/privileges/rest-reader", "execute"), let $rule := conf:get-sparql-protocol-rule() let $params := rest:process-request($rule) let $headers := eput:get-request-headers() let $method := eput:get-request-method($headers) let $env := map:map() let $params := local:validate-params($rule, $env, $params) let $body := switch ($method) case "GET" return text { fn:head((map:get($params, "query"), map:get($params, "update"))) } case "POST" return xdmp:get-request-body(eput:get-content-format($headers, $params))/node() default return fn:error((), "REST-UNSUPPORTEDMETHOD", $method) let $result := semmod:sparql-query($headers, $params, $body) let $response := semmod:results-payload($headers, $params, $result, local:sparql-callback#2) return if ($response instance of node() and $response/self::semmod:malformed-query) then (xdmp:set-response-code(400, "Malformed Query"), $response) else $response -- Graph already exists: DOMICLE

**Comment by sammefford** [comment]
Charles, what's the status on this?

**Comment by grechaw** [comment]
The change in the REST api was to make this a 400 rather than a 500.  
I'm not sure why we have a stack trace in here -- there's been back-and-forth about how verbose things should be.  But you can catch this in the Java API and make a new exception for it if you like.  More change in REST shouldn't be applied until 8.0-5 --blast radius would be significant.  I'll assign to you for consideration in Java, if you think there's something more to be done.

**Comment by sammefford** [comment]
Charles, we can discuss this after 8.0-4 ships.  But I'll be pushing for clean meaningful error messages from the REST layer.  I don't see any value in error messages coming with xquery code in them.  Do you?

**Comment by grechaw** [comment]
Well, I've spent some time with the REST error handling, reacquainting myself.  Its not currently clear to me why the message is the way it is, rather than inserted from the messages files.  Those appear to be correct.  I'll take another look later on.

**Comment by grechaw** [comment]
Hi @sammefford , @georgeajit , here's my take on this bug right now.  When first submitted, the behavior was a 500 exception, which was incorrect, because this is better understood as a 400 Bad Request, with corrective action on the client side possible.

The error message to me seems appropriate -- it's just after a big stack snippet
Graph already exists: DOMICLE

The big snippet is the entire contents of the main module at rest-api/endpoints/sparql-query.xqy

So the behavior of including the location of the eror and the messsage is consistent with other parts of the REST API.  It may be that the message is so big and ugly because its thrown directly from a main module, rather than a library module; I'm not sure.

But I think I'm recommending we ship the bug or reconsider again REST messages more carefully.  The shift to more verbose messaging happened in the 8 timeframe I think. What do you folks think?

**Comment by sammefford** [comment]
Thanks, Charles, for the clarification.  It sounds like the question here is general about REST error messages.  I've changed the bug title accordingly.  

I didn't realize the REST messages in general are including code.  I'm not seeing any benefit to including our code in our error messages.  I see value including customer code when it is the cause of the error.  But that's not the case here.  Can we change REST to only include code when it's not our code?

Also, when we do include code, would it be better to share the error message first then the code after?

**Comment by grechaw** [comment]
It may be configurable.  Let's consult the wolfpack

**Comment by grechaw** [comment]
I guess in the interest of not disturbing regression before a release, I'll propose punting on this issue for 3.0.5.

**Comment by grechaw** [comment]
@sammefford Hi Sam, I'd like to put this bug into the general backlog for Java Client API.
Paging it in and out only to triage it down the road has probably gone on too long.
I investigated the issue quite a bit, and could not really understand the behavior, but isolated it to how the error was bring thrown by the server.
I can help with investigation further, but I think you should really determine priority and perhaps decide not to fix for 9.0.1

**Comment by grechaw** [comment]
@sammefford  On the other hand, if you look at this and feel no ownership in this part of the code, you can delegate back to me and I'll try to figure out where to go from here.
